packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,748 @@
+    x2: number,
+    barHeight: number,
+    peakHeight: number
+  ) {
+    const h = ctx.canvas.height;
+    let y = h - barHeight;
+
+    ctx.drawImage(
+      this._bar,
+      0,
+      0,
+      this._bar.width,
+      h - y,
+      x,
+      y,
+      x2 - x + 1,
+      h - y
+    );
+
+    if (this._vis.peaks) {
+      const peakY = h - peakHeight;
+      ctx.drawImage(this._peak, 0, 0, 1, 1, x, peakY, x2 - x + 1, 1);
+    }
+  }
+
+  /**
+   * 🟥
+   * 🟥🟧
+   * 🟥🟧🟨
+   * 🟥🟧🟨🟩
+   */
+  paintBarLine(
+    ctx: CanvasRenderingContext2D,
+    x: number,
+    x2: number,
+    barHeight: number,
+    peakHeight: number
+  ) {
+    const h = ctx.canvas.height;
+    let y = h - barHeight;
+
+    if (!this.logged) {
+      console.log("FIXME: Line drawing is currently Fire mode!");
+      this.logged = true;
+    }
+
+    ctx.drawImage(
+      this._bar,
+      0, // sx
+      0, // sy
+      this._bar.width, // sw
+      h - y, // sh
+      x,
+      y, //  dx,dy
+      x2 - x + 1, //dw
+      h - y //dh
+    );
+
+    if (this._vis.peaks) {
+      const peakY = h - peakHeight;
+      ctx.drawImage(this._peak, 0, 0, 1, 1, x, peakY, x2 - x + 1, 1);
+    }
+  }
+}
+
+//? =============================== OSCILLOSCOPE PAINTER ===============================
+
+type PaintWavFunction = (x: number, y: number) => void;
+
+function slice1st(
+  dataArray: Uint8Array,
+  sliceWidth: number,
+  sliceNumber: number
+): number {
+  const start = sliceWidth * sliceNumber;
+  return dataArray[start];
+}
+
+export class WavePaintHandler extends VisPaintHandler {
+  _analyser: AnalyserNode;
+  _bufferLength: number;
+  _lastX: number = 0;
+  _lastY: number = 0;
+  _dataArray: Uint8Array;
+  _pixelRatio: number; // 1 or 2
+  // Off-screen canvas for drawing perfect pixel (no blurred lines)
+  _bar: HTMLCanvasElement = document.createElement("canvas");
+  _16h: HTMLCanvasElement = document.createElement("canvas"); // non-stretched
+  paintWav: PaintWavFunction;
+
+  constructor(vis: Vis) {
+    super(vis);
+    this._analyser = this._vis.analyser!;
+    this._bufferLength = this._analyser.fftSize;
+    this._dataArray = new Uint8Array(this._bufferLength);
+
+    this._16h.width = 1;
+    this._16h.height = 16;
+    this._16h.setAttribute("width", "75");
+    this._16h.setAttribute("height", "16");
+
+    //* see https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio#monitoring_screen_resolution_or_zoom_level_changes
+    this._pixelRatio = window.devicePixelRatio || 1;
+
+    if (this._vis.oscStyle === "dots") {
+      this.paintWav = this.paintWavLine.bind(this);
+    } else if (this._vis.oscStyle === "solid") {
+      // does call paintWavLine but there is a check inside
+      // that changes the oscstyle accordingly
+      this.paintWav = this.paintWavLine.bind(this);
+    } else {
+      this.paintWav = this.paintWavLine.bind(this);
+    }
+  }
+
+  prepare() {
+    if (!this._ctx) {
+      console.log("ctx not set!");
+      return;
+    }
```

Comment:
Can we just throw up above when we create the context if it's null, and then just type `this._ctx` as non-optional?